pkg/e2e/test_convergence.go

```diff
@@ -0,0 +1,320 @@
+			go func(sub apiclient.Stream) {
+				for {
+					env, err := sub.Next(ctx)
+					if err != nil {
+						if isErrClosedConnection(err) || err == context.Canceled {
+							break
+						}
+						e.log.Error("getting next", zap.Error(err))
+						break
+					}
+					if env == nil {
+						continue
+					}
+					envC <- env
+				}
+			}(sub)
+			err := subscribeExpect(envC, envs)
+			if err != nil {
+				duration := time.Since(publishStart)
+				e.log.Error("error checking subscription", zap.Error(err), zap.Duration("duration", time.Since(publishStart)))
+				e.metrics.recordSubscribeConvergence(e.ctx, name, nodeHosts[nodeIndex], "failed", duration)
+				return nil
+			}
+
+			subscribeConvergenceDuration := time.Since(publishStart)
+			e.log.Info("subscribe converged", zap.Duration("duration", subscribeConvergenceDuration), zap.String("node", nodeHosts[nodeIndex]))
+			e.metrics.recordSubscribeConvergence(e.ctx, name, nodeHosts[nodeIndex], "passed", subscribeConvergenceDuration)
+
+			return nil
+		})
+	}
+
+	// Expect that they're stored.
+	var queryGroup errgroup.Group
+	for nodeIndex, client := range clients {
+		if client == nil {
+			continue
+		}
+		client := client
+		nodeIndex := nodeIndex
+		queryGroup.Go(func() error {
+			err := expectQueryMessagesEventually(ctx, client, []string{topic}, envs)
+			if err != nil {
+				duration := time.Since(publishStart)
+				e.log.Error("error querying", zap.Error(err), zap.Duration("duration", duration))
+				e.metrics.recordQueryConvergence(e.ctx, name, nodeHosts[nodeIndex], "failed", duration)
+				return nil
+			}
+
+			duration := time.Since(publishStart)
+			e.log.Info("query converged", zap.Duration("duration", duration), zap.String("node", nodeHosts[nodeIndex]))
+			e.metrics.recordQueryConvergence(e.ctx, name, nodeHosts[nodeIndex], "passed", duration)
+
+			return nil
+		})
+	}
+
+	err = subscribeGroup.Wait()
+	if err != nil {
+		return err
+	}
+
+	err = queryGroup.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func subscribeExpect(envC chan *messagev1.Envelope, envs []*messagev1.Envelope) error {
+	receivedEnvs := []*messagev1.Envelope{}
+	waitC := time.After(5 * time.Second)
+	var done bool
+	for !done {
+		select {
+		case env := <-envC:
+			receivedEnvs = append(receivedEnvs, env)
+			if len(receivedEnvs) == len(envs) {
+				done = true
+			}
+		case <-waitC:
+			done = true
+		}
+	}
+	err := envsDiff(envs, receivedEnvs)
+	if err != nil {
+		return errors.Wrap(err, "expected subscribe envelopes")
+	}
+	return nil
+}
+
+func isErrClosedConnection(err error) bool {
+	return errors.Is(err, io.EOF) || strings.Contains(err.Error(), "closed network connection") || strings.Contains(err.Error(), "response body closed")
+}
+
+func expectQueryMessagesEventually(ctx context.Context, client apiclient.Client, contentTopics []string, expectedEnvs []*messagev1.Envelope) error {
+	timeout := 10 * time.Second
+	delay := 10 * time.Millisecond
+	// TODO: higher delay for production?
```

Comment:
Lower delay gives more precision on the convergence durations, but adds more query load. To be addressed before merging this PR.